kvs: re-work kvs.sync

chu11 opened the pull request:
The last vestige of watch in ```modules/kvs``` was the use of the "watchlist" by ```kvs.sync```.  I considered leaving the code there b/c it was still useful by ```kvs.sync```, but the implementation was inefficient.  It would continually run through the whole watchlist (err "synclist") on every setroot event.

Instead, I created a new internal lib called "kvssync", that is a very simplified version of the former "watchlist" just for ```kvs.sync```.  The "synclist" is sorted by the sequence number people are waiting on.  So everytime a setroot even occurs, you only traverse the list as far as you need to to alert syncers.

chu11 commented:
Only travis error appears to be #2015, so gotta fix that first.

chu11 commented:
rebased

garlick commented:
Nice optimization!

I had a quick spin through this and it looks good.  Coverage report will be good to see, just to make sure the new code is getting used.

grondo commented:
Would be nice to get all these kvs PRs merged in a row. If you don't mind closing issues by hand we could mergify them all, but up to you!

I leave final approval to @garlick

chu11 commented:
I saw a 85.71% diff coverage, woo-hoo!  Only notable diff coverage miss was I didn't test two kvs.sync requests with the same version number being waited on (it's a miss in the sort comparison).

Added a dumb unit test to cover that path.  Rebased & re-pushed.
